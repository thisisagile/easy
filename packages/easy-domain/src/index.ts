export * from './enums/Country';
export * from './enums/Currency';
export * from './enums/Locale';
export * from './enums/UnitOfMeasurement';
export * from './enums/UnitOfWeight';

export * from './structs/Address';
export * from './structs/Box';
export * from './structs/Dimension';
export * from './structs/Money';
export * from './structs/Name';
export * from './structs/Weight';

export * from './values/EAN';
export * from './values/Email';
export * from './values/IBAN';
export * from './values/DunsNumber';
export * from './values/PostalCode';
export * from './values/Slug';
export * from './values/Url';
