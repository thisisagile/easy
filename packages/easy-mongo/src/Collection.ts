import {
  asString,
  Condition,
  Database,
  Field, Get,
  Json,
  LogicalCondition,
  MapOptions,
  Mapper,
  mappings, ofGet,
  PropertyOptions,
  Text,
  toCondition,
  toUuid,
} from '@thisisagile/easy';
import { toMongoType } from './Utils';
import { MongoProvider } from './MongoProvider';

export class Collection extends Mapper {
  protected readonly map = {
    ...mappings,
    field: <T = unknown>(name: string, options?: PropertyOptions<T>): Field => new Field(name, options),
  };
  readonly id = this.map.field('id', { dflt: toUuid });

  constructor(options: MapOptions = { startFrom: 'source' }) {
    super(options);
  }

  get db(): Database {
    return Database.Default;
  }

  get provider(): MongoProvider {
    return new MongoProvider(this);
  }

  where = (...conditions: Get<Condition, this>[]): Json => new LogicalCondition('and', conditions.map(c => ofGet(c, this))).toJSON();

  match = (condition: Get<Condition | LogicalCondition, this>): Json => ({ $match: ofGet(condition, this).toJSON() });

  group = (...conditions: Get<Condition, this>[]): Json => new LogicalCondition('group', conditions.map(c => ofGet(c, this))).toJSON();

  google = (value: unknown): Condition => toCondition('$text', 'search', value);

  search = (key: Text): Field => this.map.field(asString(key));

  out(to: Json = {}): Json {
    return toMongoType(super.out(to));
  }
}
