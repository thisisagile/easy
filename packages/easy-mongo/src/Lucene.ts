import {
  DateTime,
  entries,
  Func,
  ifDefined,
  ifNotEmpty,
  ifTrue,
  isDefined,
  isEmptyObject,
  isFunction,
  List,
  on,
  OneOrMore,
  Optional,
  RequireAtLeastOne,
  toArray,
  toList,
} from '@thisisagile/easy';
import { toMongoType } from './Utils';

type FuzzyOptions = {
  maxEdits: number;
  prefixLength: number;
  maxExpansions: number;
};

export type Facet =
  | { path: string; type: 'string'; numBuckets: number }
  | { path: string; type: 'number'; boundaries: number[]; default?: string }
  | { path: string; type: 'date'; boundaries: Date[]; default?: string };

export type Operator = Func<Optional<object>, string>;
export type Clause = object | Operator;
export type Clauses = Record<string, Clause>;
export type SearchDefinition = Record<
  string,
  (
    v: string | number,
    q?: Record<string, string | number>
  ) => RequireAtLeastOne<{
    should?: Clauses;
    filter?: Clauses;
    must?: Clauses;
    mustNot?: Clauses;
    sort?: Record<string, 1 | -1>;
    facet?: Facet;
  }>
>;

type Compound = {
  must: OneOrMore<Clauses>;
  should: OneOrMore<Clauses>;
  mustNot: OneOrMore<Clauses>;
  filter: OneOrMore<Clauses>;
};

const should = (query: Record<string, string | number>, def: SearchDefinition): Clauses[] => entries(query).mapDefined(([k, v]) => def[k]?.(v, query)?.should);
const must = (query: Record<string, string | number>, def: SearchDefinition): Clauses[] => entries(query).mapDefined(([k, v]) => def[k]?.(v, query)?.must);
const mustNot = (query: Record<string, string | number>, def: SearchDefinition): Clauses[] =>
  entries(query).mapDefined(([k, v]) => def[k]?.(v, query)?.mustNot);
const filter = (query: Record<string, string | number>, def: SearchDefinition): Clauses[] => entries(query).mapDefined(([k, v]) => def[k]?.(v, query)?.filter);

export const lucene = {
  clause: (c: Clauses) => entries(c).reduce((res, [k, v]) => res.add(isFunction(v) ? v(k) : v), toList()),
  clauses: (cs: OneOrMore<Clauses>) => toArray(cs).flatMap(c => lucene.clause(c)),
  compound: (query: Record<string, string | number>, def: SearchDefinition, wildcard = true): Partial<Compound> =>
    ifNotEmpty(
      entries({
        should: should(query, def),
        filter: filter(query, def),
        mustNot: mustNot(query, def),
        must: must(query, def),
      }).filter(([_, v]) => v.length > 0),
      e => e.reduce((res, [k, v]) => on(res, r => (r[k] = lucene.clauses(v))), should(query, def).length > 0 ? { minimumShouldMatch: 1 } : ({} as any)),
      () =>
        ifTrue(wildcard, () => ({
          should: lucene.clauses([{ wildcard: lucene.wildcard() }]),
          minimumShouldMatch: 0,
        }))
    ),
  search: (c: Partial<Compound>, index?: string) => ({
    $search: {
      ...ifDefined(index, { index }),
      compound: entries(c).reduce((res, [k, v]) => on(res, r => (r[k] = lucene.clauses(v))), {} as any),
    },
  }),
  searchWithDef: (query: Record<string, string | number>, options: SearchDefinition, count: 'total' | 'lowerBound' = 'total', index?: string) => {
    const sort = entries(query)
      .mapDefined(([k, v]) => options[k]?.(v, query)?.sort)
      .first();
    return {
      $search: {
        ...ifDefined(index, { index }),
        compound: lucene.compound(query, options),
        ...ifDefined(sort, { sort }),
        count: { type: count },
      },
    };
  },
  searchMeta: (query: Record<string, string | number>, def: SearchDefinition, count: 'total' | 'lowerBound' = 'total', index?: string) => ({
    $searchMeta: {
      ...ifDefined(index, { index }),
      ...ifTrue(
        !isEmptyObject(lucene.facets(def)),
        {
          facet: {
            operator: {
              compound: lucene.compound(query, def),
            },
            facets: lucene.facets(def),
          },
        },
        { compound: lucene.compound(query, def) }
      ),
      count: { type: count },
    },
  }),
  exists: (): Operator => (path: string) => ({ exists: { path } }),
  text:
    (value?: OneOrMore<unknown>, fuzzy?: Partial<FuzzyOptions>): Operator =>
    (path: string) =>
      ifDefined(value, v => ({
        text: {
          path: path === 'wildcard' ? { wildcard: '*' } : path,
          query: v,
          ...ifDefined(fuzzy, { fuzzy }),
        },
      })),
  wildcard:
    (value?: OneOrMore<unknown>, allowAnalyzedField = true): Operator =>
    (path: string) => ({
      wildcard: {
        path: path === 'wildcard' ? { wildcard: '*' } : path,
        query: ifDefined(value, value, '*'),
        allowAnalyzedField,
      },
    }),
  lt:
    (value: unknown): Operator =>
    (path: string) =>
      ifDefined(value, lt => ({ range: { path, lt } })),
  lte:
    (value: unknown): Operator =>
    (path: string) =>
      ifDefined(value, lte => ({ range: { path, lte } })),
  gt:
    (value: unknown): Operator =>
    (path: string) =>
      ifDefined(value, gt => ({ range: { path, gt } })),
  gte:
    (value: unknown): Operator =>
    (path: string) =>
      ifDefined(value, gte => ({ range: { path, gte } })),
  after: (date: unknown): Operator => lucene.gte(toMongoType(date)),
  before: (date: unknown): Operator => lucene.lt(toMongoType(date)),
  between:
    (after: unknown, before: unknown, includeLimit?: boolean): Operator =>
    (path: string) => {
      const upperLimit = includeLimit ? { lte: toMongoType(before) } : { lt: toMongoType(before) };
      return {
        range: {
          path,
          gte: toMongoType(after),
          ...upperLimit,
        },
      };
    },
  facets: (def: SearchDefinition) =>
    entries(def)
      .filter(([k, v]) => isDefined(v(k)?.facet))
      .map(([k, v]) => ({ [k]: v(k)?.facet }))
      .reduce((acc, v) => ({ ...acc, ...v }), {}),
  facet: {
    string: (path: string, numBuckets = 1000): Facet => ({
      type: 'string',
      path,
      numBuckets,
    }),
    number: (path: string, boundaries: List<number>, alt?: string): Facet => ({
      type: 'number',
      path,
      boundaries,
      ...ifDefined(alt, a => ({ default: a })),
    }),
    date: (path: string, boundaries: List<DateTime>, alt?: string): Facet => ({
      type: 'date',
      path,
      boundaries: boundaries.mapDefined(b => b.toDate()),
      ...ifDefined(alt, a => ({ default: a })),
    }),
  },
};
