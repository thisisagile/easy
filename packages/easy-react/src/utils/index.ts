export * from './Children';
export * from './Hooks';
export * from './useOnce';
