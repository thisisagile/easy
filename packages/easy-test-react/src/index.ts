export * from './ElementTester';
export * from './Tester';
export * from './waitForRender';
