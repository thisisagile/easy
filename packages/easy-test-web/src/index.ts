export * from './PlaywrightElement';
export * from './PlaywrightTester';
export * from './PuppeteerElement';
export * from './PuppeteerTester';
export * from './TestElement';
export * from './Tester';
