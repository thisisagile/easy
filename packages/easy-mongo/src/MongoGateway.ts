import { MongoProvider } from './MongoProvider';
import { asJson, Condition, Gateway, Id, ifDefined, isDefined, Json, JsonValue, List, LogicalCondition } from '@thisisagile/easy';
import { Collection } from './Collection';

export class MongoGateway implements Gateway {
  constructor(readonly collection: Collection, readonly provider: MongoProvider = new collection.db.provider(collection)) {}

  all(): Promise<List<Json>> {
    return this.provider.all().then(l => l.map(j => this.collection.in(j)));
  }

  byId(id: Id): Promise<Json | undefined> {
    return this.provider.byId(id).then(j => ifDefined(j, this.collection.in(j)));
  }

  by(key: string, value: JsonValue): Promise<List<Json>> {
    return this.provider.by(key, value).then(l => l.map(j => this.collection.in(j)));
  }

  find(q: JsonValue | Condition | LogicalCondition): Promise<List<Json>> {
    return this.provider.find(asJson(q)).then(l => l.map(j => this.collection.in(j)));
  }

  search(q: JsonValue): Promise<List<Json>> {
    return this.find(this.collection.google(q));
  }

  exists(id: Id): Promise<boolean> {
    return this.provider.byId(id).then(i => isDefined(i));
  }

  add(item: Json): Promise<Json> {
    return this.provider.add(this.collection.out(item)).then(j => this.collection.in(j));
  }

  update(item: Json): Promise<Json> {
    return this.provider.update(this.collection.out(item)).then(j => this.collection.in(j));
  }

  remove(id: Id): Promise<boolean> {
    return this.provider.remove(id);
  }
}
