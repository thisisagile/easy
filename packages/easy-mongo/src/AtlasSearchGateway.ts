import { stages } from './Stages';
import { MongoGateway } from './MongoGateway';
import { Collection } from './Collection';
import { MongoProvider } from './MongoProvider';
import { lucene, SearchDefinition } from './Lucene';
import { asNumber, entries, Filter, Json, PageList, text, toPageList, tuple2 } from '@thisisagile/easy';
import { Filter as MongoFilter} from '../src/MongoProvider';

const { skip, take } = stages.skip;
const { replaceWith } = stages.replaceWith;
const { searchWithDef, searchMeta, facets } = lucene;

type FacetResult = { _id: string; count: number };
type Value = { label: string; value: string; count: number };
export const toFilters = (facets: unknown): Filter[] =>
  entries(facets as Record<string, unknown>).map(([k, fs]) => ({
    label: text(k).title.toString(),
    field: k,
    values: (fs as []).map((f: FacetResult) => toValue(f)),
  }));

const toValue = (f: FacetResult): Value => ({ label: f._id ?? 'unknown', value: f._id ?? 'unknown', count: f.count });

export class AtlasSearchGateway extends MongoGateway {
  constructor(
    collection: Collection,
    readonly searchDef: SearchDefinition,
    readonly sortDef: Record<string, Record<string, 1 | -1>> = {},
    provider: MongoProvider = collection.provider
  ) {
    super(collection, provider);
  }

  query(query: Record<keyof typeof this.searchDef, string | number>, additionalStages: MongoFilter[] = []): Promise<PageList<Json>> {
    return tuple2(
      this.aggregate(
        searchWithDef(query, this.searchDef),
        skip({ skip: (query?.skip as number) ?? 0 }),
        take({ take: (query?.take as number) ?? 250 }),
        ...additionalStages
      ),
      this.aggregate(
        searchMeta(query, this.searchDef),
        replaceWith({
          total: '$count.total',
          facets: Object.keys(facets(this.searchDef)).reduce((acc, k) => ({ ...acc, [k]: `$facet.${k}.buckets` }), {}),
        })
      )
    )
      .then(([data, meta]) => ({ data, meta: meta.first() }))
      .then(({ data, meta }) =>
        toPageList<Json>(data, {
          total: asNumber(meta?.total, 0),
          skip: asNumber(query?.skip, 0),
          take: asNumber(query?.take, 250),
          sorts: Object.keys(this.sortDef),
          filters: toFilters(meta.facets),
        })
      );
  }
}
