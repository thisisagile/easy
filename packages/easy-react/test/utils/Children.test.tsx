import React, { ReactNode } from 'react';
import '@thisisagile/easy-test';
import { rendersWait } from '@thisisagile/easy-test-react';
import { toChildren } from '../../src';

const Things = ({ children }: { children?: ReactNode }) => <div>{toChildren(children).length}</div>;

describe('Children', () => {
  test('renders correctly without children.', async () => {
    const { container, byText } = await rendersWait(<Things></Things>);
    expect(container).toMatchSnapshot();
    expect(byText('0')).toBeDefined();
  });

  test('renders correctly with children.', async () => {
    const { container, byText } = await rendersWait(<Things>Text</Things>);
    expect(container).toMatchSnapshot();
    expect(byText('1')).toBeDefined();
  });

  test('renders correctly with multiple children.', async () => {
    const { container, byText } = await rendersWait(
      <Things>
        <span>Hoi</span>
        <div>Hallo</div>
      </Things>
    );
    expect(container).toMatchSnapshot();
    expect(byText('2')).toBeDefined();
  });
});
