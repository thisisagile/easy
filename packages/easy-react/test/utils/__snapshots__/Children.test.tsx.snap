// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Children renders correctly with children. 1`] = `
<div>
  <div>
    1
  </div>
</div>
`;

exports[`Children renders correctly with multiple children. 1`] = `
<div>
  <div>
    2
  </div>
</div>
`;

exports[`Children renders correctly without children. 1`] = `
<div>
  <div>
    0
  </div>
</div>
`;

exports[`Children renders correctly. 1`] = `
<div>
  <div>
    0
  </div>
</div>
`;
